src/functional/merge.js

```diff
@@ -69,10 +72,22 @@ export function mergeWithSources(collection, sources, merger) {
 function deepMergerWith(merger) {
   function deepMerger(oldValue, newValue, key) {
     return isDataStructure(oldValue) && isDataStructure(newValue)
-      ? mergeWithSources(oldValue, [newValue], deepMerger)
+      ? areMergeable(oldValue, newValue)
+        ? mergeWithSources(oldValue, [newValue], deepMerger)
+        : newValue
       : merger
       ? merger(oldValue, newValue, key)
       : newValue;
   }
   return deepMerger;
 }
+
+/**
+ * The data structures are considered not to be mergeable if one of them is indexed and the other is keyed.
+ */
+function areMergeable(oldDataStructure, newDataStructure) {
+  return !(
+    (isIndexed(Seq(oldDataStructure)) && isKeyed(Seq(newDataStructure))) ||
+    (isKeyed(Seq(oldDataStructure)) && isIndexed(Seq(newDataStructure)))
+  );
```

Comment:
How does this handle `Set` types?

Should this be something like

```suggestion
  return (
    (isIndexed(Seq(oldDataStructure)) === isIndexed(Seq(newDataStructure))) &&
    (isKeyed(Seq(oldDataStructure)) === isKeyed(Seq(newDataStructure)))
  );
```

Reply:
> How does this handle `Set` types?

It doesn't. 😄

I'm open to that change, I was just erring on the side of being more permissive in case there were cases I wasn't considering. How *do* we want to handle sets? Should they be mergeable with lists, but not maps?

Reply:
If we're optimizing for "easy to explain" then I think we probably want to follow the model here. Lists only merge into Lists, Maps into Maps, Sets into Sets. What do you think?

Reply:
Yeah, that seems the most straightforward rule. Although Maps into Maps and Sets or Lists into Sets or Lists could also be fairly easy to explain. I'm trying to think what people would be expecting and I'm having a hard time guessing since I can see it going either way as far as desired behavior. I guess the nice thing about keeping it more strict is that it will call the merger function and so that gives the user the ability to easily override it.

Reply:
Updated.

Reply:
>  I guess the nice thing about keeping it more strict is that it will call the merger function and so that gives the user the ability to easily override it. The other way it would be much harder to override.

That really convinces me!